src/amo/withExperiment.js

```diff
@@ -175,15 +181,79 @@ export const withExperiment =
         WrappedComponent,
       )})`;
 
-      componentDidMount() {
-        const { _getVariant, _isExperimentEnabled, cookies, id, variants } =
-          this.props;
-
+      initialExperimentSetup(props) {
+        const {
+          _getVariant,
+          _isExperimentEnabled,
+          dispatch,
+          id,
+          storedVariant,
+          variants,
+        } = props;
+
+        let { variant } = this;
         const isEnabled = _isExperimentEnabled({ _config, id });
         const registeredExperiments = this.getExperiments();
         const experimentInCookie = this.cookieIncludesExperiment(
           registeredExperiments,
         );
+        const addExperimentToCookie = !experimentInCookie;
+        const variantFromStore = storedVariant;
+
+        if (isEnabled && !variant) {
+          // Use the variant in the cookie if one exists, otherwise use the
+          // variant from the Redux store.
+          if (experimentInCookie) {
+            variant = registeredExperiments[id];
+          } else if (variantFromStore) {
+            variant = variantFromStore.variant;
+          }
+
+          // Do we need to store the variant in the cookie?
+          if (addExperimentToCookie) {
+            // Determine the variant if we don't already have one.
+            variant = variant || _getVariant({ variants }).id;
+
+            // If we are on the server, we need to store the variant in the Redux
+            // store.
+            if (_config.get('server')) {
```

Comment:
Why on the server only?

Reply:
We don't need it on the client, because the cookie will be set and the enrollment event will get dispatched during the render, so there is no need to store it. I don't think we should dispatch an action if it's not needed.

Reply:
heh ok, I see. We need more comments if we start to have distinct behaviors on the server and on the client (I think dispatching an action isn't that bad if that keeps things a bit easier and more consistent).

Reply:
Just trying to understand what you'd prefer here. Are you saying you'd prefer to dispatch an unnecessary action on the client, if it means we don't have different code paths for the client vs. the server?

Reply:
> Are you saying you'd prefer to dispatch an unnecessary action on the client, if it means we don't have different code paths for the client vs. the server?

Yes, unless you disagree and in this case, we'll need comments :)

Reply:
I don't *strongly* disagree, so I will change it.